lib/src/annotate.rs

```diff
@@ -0,0 +1,421 @@
+    /// If there is already a file for a given commit, it is a no-op.
+    fn load_file_into_cache(
+        &mut self,
+        store: &Store,
+        commit_id: &CommitId,
+        file_path: &RepoPath,
+        file_id: &FileId,
+    ) -> BackendResult<()> {
+        if self.file_cache.contains_key(commit_id) {
+            return Ok(());
+        }
+
+        let file_contents = get_file_contents(store, file_path, file_id)?;
+        self.file_cache.insert(commit_id.clone(), file_contents);
+
+        Ok(())
+    }
+
+    /// retrieves a file from the cache for a specific commit. If the file isn't
+    /// found, it panics. Make sure to call load_file_into_cache first.
+    fn get_file_from_cache(&self, commit_id: &CommitId) -> &[u8] {
+        self.file_cache.get(commit_id).unwrap()
+    }
+}
+
+/// Get line by line annotations for a specific file path in the repo.
+pub fn get_annotation_for_file(
+    file_path: &RepoPath,
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+) -> Result<AnnotateResults, AnnotateError> {
+    let store = repo.store();
+
+    let current_commit = store.get_commit(starting_commit_id)?;
+    let current_tree = current_commit.tree()?;
+    let original_file_id = get_file_id(current_commit.id(), &current_tree, file_path)?;
+    if original_file_id.is_none() {
+        return Err(AnnotateError::FileNotFound(
+            file_path.as_internal_file_string().to_string(),
+        ));
+    }
+    let original_contents = get_file_contents(repo.store(), file_path, &original_file_id.unwrap())?;
+    let num_lines = original_contents.split_inclusive(|b| *b == b'\n').count();
+    let mut partial_results = PartialResults::new(starting_commit_id, num_lines);
+
+    process_commits(
+        &mut partial_results,
+        repo,
+        current_commit.id(),
+        file_path,
+        num_lines,
+    )?;
+
+    Ok(partial_results.convert_to_results(&original_contents))
+}
+
+/// Starting at the starting commit, compute changes at that commit, updating
+/// the mappings. So long as there are mappings left in local_line_map, we
+/// continue. Once local_line_map is empty, we've found sources for each line
+/// and exit.
+fn process_commits(
+    results: &mut PartialResults,
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+    file_name: &RepoPath,
+    num_lines: usize,
+) -> Result<(), AnnotateError> {
+    let predicate = RevsetFilterPredicate::File(FilesetExpression::file_path(file_name.to_owned()));
+    let revset = RevsetExpression::commit(starting_commit_id.clone())
+        .ancestors()
+        .filtered(predicate)
+        .evaluate_programmatic(repo)
+        .unwrap();
+    let mut is_first = true;
+
+    for (cid, edge_list) in revset.iter_graph() {
+        if is_first {
+            results.swap_full_commit_id(starting_commit_id, &cid);
+            is_first = false;
+        }
+        process_commit(results, repo, file_name, &cid, &edge_list)?;
```

Comment:
Perhaps, the starting commit has to be included in the graph (i.e. `ancestors(start) & (file(path) | start)`?

The starting-commit lines can be forwarded to multiple parents if there are a merge commit in between. `file(path)` won't include a merge commit if the commit is empty relative to auto-merge parents.

Reply:
Good idea, implemented